### PR TITLE
Make no_log=False override the no_log warnings.

### LINKED ISSUE #49465: password/pass no_log
##### SUMMARY
I'm working on custom module for ansible, in this module we have several vars which are include *password* key word. For all those var I get warring for no_log usage however they no need no_log to be set to true. Same is If I change it form *password* to *pass*

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
module no_log

##### ANSIBLE VERSION
```
ansible 2.8.0.dev0
  config file = /root/.ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible-2.8.0.dev0-py2.7.egg/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Jul 13 2018, 13:06:57) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

##### CONFIGURATION
```
HOST_KEY_CHECKING(/root/.ansible.cfg) = False
```

##### OS / ENVIRONMENT
Linux **** 3.10.0-862.11.6.el7.x86_64 #1 SMP Tue Aug 14 21:49:04 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

##### STEPS TO REPRODUCE
```yaml

  tasks:
    - name: vCenter localos users
      vmware_vcenter_localos_users:
        hostname: '{{ vcenter_hostname }}'
        username: '{{ vcenter_username }}'
        password: '{{ vcenter_password }}'
        validate_certs:   '{{ valid_cert }}'
        localos_username: '{{ item.value.username }}'
        localos_password: '{{ item.value.password }}'
        localos_state:    '{{ item.value.state }}'
        localos_role:     '{{ item.value.role }}'
        localos_fullname:     '{{ item.value.fullname }}'
        localos_email:  '{{ item.value.email }}'
        localos_warn_days_before_password_expiration: '{{ item.value.warn_days_before_password_expiration }}'
      loop: "{{ localos_users|dict2items }}"


```

```
def main():
    argument_spec = VmwareRestClient.vmware_client_argument_spec()
    argument_spec.update(localos_username=dict(type='str', required=True),
                         localos_password=dict(type='str', required=True, no_log=True),
                         localos_state=dict(type='str', choices=['present', 'absent'], default='present', required=False),
                         localos_role=dict(type='list', required=False),
                         localos_email=dict(type='str', required=False, default=None),
                         localos_fullname=dict(type='str', required=False, default=None),
                         localos_enabled=dict(type='bool', required=False, default=True),
                         localos_password_expires=dict(type='bool', required=False, default=False),
                         localos_password_expires_at=dict(type='str', required=False, default=None),
                         localos_inactive_after_password_expiration=dict(type='bool', required=False, default=False),
                         localos_days_after_password_expiration=dict(type='int', required=False, default=None),
                         localos_min_days_between_password_change=dict(type='int', required=False, default=None),
                         localos_max_days_between_password_change=dict(type='int', required=False, default=None),
                         localos_warn_days_before_password_expiration=dict(type='int', required=False, default=None))

    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)

```

##### EXPECTED RESULTS
We should not see no_log warnings for different vars which are include *password* or *pass* in name (we don't need to specify no_log=True for example:  **localos_min_days_between_password_change**) etc.

##### ACTUAL RESULTS

```
localos_password is set with no_log=True which is ok but still we get warnings for all other vars.

 [WARNING]: Module did not set no_log for localos_min_days_between_password_change

 [WARNING]: Module did not set no_log for localos_password_expires_at

 [WARNING]: Module did not set no_log for localos_warn_days_before_password_expiration

 [WARNING]: Module did not set no_log for localos_max_days_between_password_change

 [WARNING]: Module did not set no_log for localos_days_after_password_expiration


```

### LINKED ISSUE #64656: Honor explicitly marking a parameter as not sensitive.
##### SUMMARY

The way that the `no_log` parameter is currently implemented in `AnsibleModule._log_invocation` is overly simplistic, and appears to violate the principle of least astonishment due to its overly brief documentation failing to mention the corner cases for which it appears to have been written. As `no_log` is currently implemented, it will:

* Not log a parameter value if `no_log` is truthy.
* Not log a parameter value and complain if the parameter name matches `PASSWORD_MATCH` and is neither a `bool` or enum.
* Otherwise, log the parameter.

On the surface (as a module writer), the documentation (see: [Protecting sensitive data with `no_log`](https://docs.ansible.com/ansible/devel/reference_appendices/logging.html?highlight=no_log#protecting-sensitive-data-with-no-log)) states that by passing `no_log=True` you are explicitly marking a parameter as sensitive. This makes sense. It would then follow that by passing `no_log=False` you are explicitly marking a parameter as not sensitive. This is where the current implementation fails. There is currently no way (that I can see) for a module author that has properly developed and tested a module to avoid a warning regarding the `no_log` parameter if the parameter matches `PASSWORD_MATCH` (at least, not that I can see in the documentation or HEAD.)

I certainly understand the need to protect users from the whims of careless third party module authors (some of which may have a tenuous grasp on Python at best), but for those of us that write and maintain our own modules professionally it seems like a rather strange choice to provide no avenue for a knowledgeable module author to say, "no, pass_interval really is not a password, and no, there is no other name that would work without causing excessive end-user confusion".

The feature request is basically: can `AnsibleModule._log_invocation` be adjusted to actually honor `no_log=False` as the opposite of `no_log=True`, which it currently does not. I do not believe it would be difficult to implement while maintaining the current behavior where a module author that does not specify either way would be caught by `PASSWORD_MATCH`.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

lib/ansible/module_utils/basic.py

##### ADDITIONAL INFORMATION

This change would not be user facing (typically) unless you're a module author. It is needed because, at the current time, there is no way to mark a parameter as explicitly not sensitive (as `no_log=False` is not being honored effectively) and avoid erroneous messages in the logs. Some authors have been getting around this by choosing more obscure names that diverge from current documentation for their parameters, but you're basically hiding from the problem at that point, and confusing end-users.

Duplicate of a proposed solution to https://github.com/ansible/ansible/issues/49465

### PR DESCRIPTION
**Pull request by nnhanndakara**
##### SUMMARY

As AnsibleModule._log_invocation is currently implemented, any parameter
with a name that matches PASSWORD_MATCH triggers the no_log warning as a
precaution against parameters that may contain sensitive data, but have not
been marked as sensitive by the module author.

This patch would allow module authors to explicitly mark the aforementioned
parameters as not sensitive thereby bypassing an erroneous warning message,
while still catching parameters which have not been marked at all by the
author.

Fixes #49465
Fixes #64656

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

lib/ansible/module_utils/basic.py

##### ADDITIONAL INFORMATION

See #49465 and #64656.

Before (with pass_interval set to no_log=False):
```paste below
% ansible -m mac_domain_membership -a "dns_domain_name=x domain_admin_password=x domain_admin_user=x" --check xxxxxxxx-xxxxxx
[WARNING]: Module did not set no_log for pass_interval

xxxxxxxx-xxxxxx | SUCCESS => {
*snip*
}
```
After (with pass_interval set to no_log=False):
```paste below
% ansible -m mac_domain_membership -a "dns_domain_name=x domain_admin_password=x domain_admin_user=x" --check xxxxxxxx-xxxxxx
xxxxxxxx-xxxxxx | SUCCESS => {
*snip*
}
```

### CODE REVIEW

**Comment by kaorihinata** [comment]
It's worth mentioning at this point that the way that `AnsibleModule._log_invocation` makes decisions about `no_log` parameters does not match how `AnsibleModule._handle_no_log_values` (and underlying `list_no_log_values()`) handles them. This is true both before and after this change, meaning that sanitizing output for parameters that match `PASSWORD_MATCH` is still just as broken as it always has been. If a more complete solution is desired, then by all means, we can close this pull request.

**Review by bcoca — APPROVED** [review]
The password matching will never be perfect, this is mostly meant as a warning/fallback, as such this change does do the right thing giving the developer control. 
The warning will work in 'absence' of information and only trigger when there is lack of intended behaviour. Good change.

**Comment by kaorihinata** [comment]
That's fine. I just figured I would mention it.

I was going to go ahead and fix the test failure by updating the core user module to take advantage of this change. It looks like the extra conditions that had been stapled onto the `PASSWORD_MATCH` check were added specifically for situations like this, but having module developers explicitly acknowledge it would replace such behavior. Should I go ahead and push it?

**Comment by kaorihinata** [comment]
As far as I'm aware all tests related to the changed code are passing. The one test that is failing is failing due to instability (azure_rm_networkinterface, in this case). I do not believe I have the ability to rerun the test from here, and the test completed successfully when it was rerun with maximum verbosity.

**Comment by Akasurde** [comment]
@bcoca Are we good at merging this? Thanks,

File: changelogs/fragments/64733-make-no_log-false-override-no_log-warnings.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - "core: Make no_log=False override the no_log warnings - fixes #49465 and #64656"
```
Comment (samdoran) [inline changelogs/fragments/64733-make-no_log-false-override-no_log-warnings.yml]:
```suggestion
  - "make ``no_log=False`` on a module option silence the ``no_log`` warning (https://github.com/ansible/ansible/issues/49465 https://github.com/ansible/ansible/issues/64656)"
```

File: lib/ansible/modules/system/user.py [hunk 2]
```diff
@@ -2859,9 +2859,9 @@ def main():
             ssh_key_file=dict(type='path'),
             ssh_key_comment=dict(type='str', default=ssh_defaults['comment']),
             ssh_key_passphrase=dict(type='str', no_log=True),
-            update_password=dict(type='str', default='always', choices=['always', 'on_create']),
+            update_password=dict(type='str', default='always', choices=['always', 'on_create'], no_log=False),
```
Comment (samdoran) [inline lib/ansible/modules/system/user.py]: These changes to the `user` module are good, but should be done in a separate PR.

**Review by samdoran — CHANGES_REQUESTED** [review]
This is a very good change. Please add unit tests to the [existing argument spect tests](https://github.com/ansible/ansible/blob/devel/test/units/module_utils/basic/test_argument_spec.py) for this change.

**Comment by samdoran** [comment]
We should also update the docs explaining this can be set to `True` _or_ `False` for a module option now. That will be very helpful information for module authors.

File: lib/ansible/modules/system/user.py (hunk 2 above)
Comment (nnhanndakara) [reply lib/ansible/modules/system/user.py]: Agreed, but just to confirm, the only reason I made this change was that, as the current tests are written for the `user` module, the existence of any warning fails the tests, which fails CI. Is this okay?

Comment (samdoran) [reply lib/ansible/modules/system/user.py]: We'll need to merge the user module PR before this one. Once that PR is merged, tests for this PR will pass.

Comment (nnhanndakara) [reply lib/ansible/modules/system/user.py]: Pull request #66219 has been created. Is it okay to mark this as resolved?

Also, please let me know if anything needs to be done or redone. I'm not the most familiar with GitHub's tools yet and I'm already regretting not creating a branch for this pull request.

Comment (samdoran) [reply lib/ansible/modules/system/user.py]: > Pull request #66219 has been created. Is it okay to mark this as resolved?

Yup.

> Also, please let me know if anything needs to be done or redone.

You can remove the changes to the `user` module from this PR.

> I'm not the most familiar with GitHub's tools yet and I'm already regretting not creating a branch for this pull request.

Creating new branches in your fork for a PR isn't required but it does make it easier to segment work.

**Comment by samdoran** [comment]
Re-running tests now that #66219 is merged.

**Comment by samdoran** [comment]
Changelog and code change looks good. We still need unit tests and documentation updates. Let me know if you need help with that.

**Comment by kaorihinata** [comment]
For the documentation, `docs/docsite/rst/dev_guide/developing_modules_documenting.rst` and `docs/docsite/rst/dev_guide/developing_program_flow_modules.rst` seem to be the only documents that will be affected. Does this sound correct?

I noticed that the C# module for Powershell implements `no_log` in some form, but that it seems to operate differently than the Python module.

Regarding the tests, should I just add a VALID_SPECS case for each value of no_log?

**Comment by kaorihinata** [comment]
Now that I look at it more closely, I don't think I'll be able to get away with that.

Additionally, it would appear that AWS is having a moment.

**Comment by bcoca** [comment]
checked with windows folk, they don't have a matching feature so no need to update the powershell common code.

**Comment by samdoran** [comment]
> For the documentation, docs/docsite/rst/dev_guide/developing_modules_documenting.rst and docs/docsite/rst/dev_guide/developing_program_flow_modules.rst seem to be the only documents that will be affected. Does this sound correct?

Yes.

> I noticed that the C# module for Powershell implements no_log in some form, but that it seems to operate differently than the Python module.

The C# implementation may already match the behavior in this PR. I don't see anything in `Ansible.Basic.cs` that is warning based on a regexp match of the module parameter names. Maybe @jborean93 or @nitzmahone can add further clarification.

> Regarding the tests, should I just add a VALID_SPECS case for each value of no_log?

It's probably easiest to add new test(s) specifically for testing the various `no_log` settings. Here's a basic example:

```python
@pytest.mark.parametrize('stdin', [{'ok_password': 'notsecret'}], indirect=['stdin'])
def test_no_log_false_no_warning(stdin, capfd):
    spec = {
        'ok_password': {'no_log': False},
    }
    am = basic.AnsibleModule(spec)
    assert len(am._warnings) == 0
```

**Comment by samdoran** [comment]
> Additionally, it would appear that AWS is having a moment.

That happens a lot. 😄

**Comment by bcoca** [comment]
@samdoran read my post, i asked them already during windows irc meeting

**Comment by kaorihinata** [comment]
> That happens a lot. 😄

Serves me right for watching the build status. I'll have the documentation in some time this evening.

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst [hunk 3]
```diff
@@ -119,7 +119,8 @@ After the shebang, the UTF-8 coding, the copyright line, the license, and the ``
 Module documentation should briefly and accurately define what each module and option does, and how it works with others in the underlying system. Documentation should be written for broad audience--readable both by experts and non-experts.
     * Descriptions should always start with a capital letter and end with a full stop. Consistency always helps.
     * Verify that arguments in doc and module spec dict are identical.
-    * For password / secret arguments no_log=True should be set.
+    * For password / secret arguments ``no_log=True`` should be set.
+    * For arguments that are redacted automatically (e.g. "password_length"), but are not password / secret arguments, set ``no_log=False``.
```
Comment (samdoran) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]:
```suggestion
    * For arguments that seem to contain sensitive information but **do not** contain secrets, such as "password_length", set ``no_log=False`` to disable the warning message.
```

File: docs/docsite/rst/dev_guide/developing_program_flow_modules.rst [hunk 4]
```diff
@@ -615,7 +615,10 @@ required
 no_log
 """"""
 
-``no_log`` indicates that the value of the argument should not be logged or displayed.
+``no_log`` accepts a boolean, either ``True`` or ``False`` that indicates explicitly whether or not the argument value should appear redacted in logs and output.
```
Comment (samdoran) [inline docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]:
```suggestion
``no_log`` accepts a boolean, either ``True`` or ``False``, that indicates explicitly whether or not the argument value should be masked in logs and output.
```

File: docs/docsite/rst/dev_guide/developing_program_flow_modules.rst [hunk 5]
```diff
@@ -615,7 +615,10 @@ required
...
+
+.. note::
+   In the absence of ``no_log``, the argument value will only appear redacted in logs and output if the argument name appears to indicate that the argument value is a password or passphrase (e.g. "admin_password"). To ensure that an argument value always appears in logs and output, and avoid misleading warnings about ``no_log``, set ``no_log`` explicitly to ``False``.
```
Comment (samdoran) [inline docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]:
```suggestion
   In the absence of ``no_log``, the argument value will only be masked in logs and output if the argument name appears to indicate that the argument value is a password or passphrase (such as "admin_password"). To ensure that an argument value always appears in logs and output, and to disable the warning about ``no_log``, set ``no_log`` explicitly to ``False``.
```

**Review by samdoran — CHANGES_REQUESTED** [review]
Some minor changes to the wording in the documentation.

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst (hunk 3 above)
Comment (samdoran) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: We avoid using Latin abbreviations like 'e.g.' that are less familiar to those for whom English is not their native language.

**Comment by kaorihinata** [comment]
Masked! That's much better. I don't know what it was last night but I was drawing a blank about what to put there. I'll remember the latin policy in the future.

Regarding the comma, I had assumed a comma would be appropriate there but the description for `required` above `no_log` excludes it. Should one be insert there as well for consistency, or should that be reserved for another pull request specifically targeting documentation corrections?

**Comment by samdoran** [comment]
We don't need to worry about the comma. It's ok how it is for now.

File: docs/docsite/rst/dev_guide/developing_program_flow_modules.rst [hunk 6]
```diff
@@ -615,7 +615,10 @@ required
 no_log
 """"""
 
-``no_log`` indicates that the value of the argument should not be logged or displayed.
+``no_log`` accepts a boolean, either ``True`` or ``False``, that indicates explicitly whether or not the argument value should be masked in logs and output.
+
+.. note::
+   In the absence of ``no_log``, the argument value will only be masked in logs and output if the argument name appears to indicate that the argument value is a password or passphrase (such as "admin_password"). To ensure that an argument value always appears in logs and output, and to disable the warning about ``no_log``, set ``no_log`` explicitly to ``False``.
```
Comment (samdoran) [inline docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: Actually, I don't think this is accurate. The value is not masked if the parameter name contains something like 'password' — only a warning is issued.
```suggestion
   In the absence of ``no_log``, if the parameter name appears to indicate that the argument value is a password or passphrase (such as "admin_password"), a warning will be shown and the value will be masked in logs but **not** output. To disable the warning and masking for parameters that do not contain sensitive information, set ``no_log`` to ``False``.
```

Comment (nnhanndakara) [reply docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: I'm not sure if it's that simple. `log_args[param]` does get initialized to `"NOT_LOGGING_PASSWORD"`, but I haven't finished looking at where that value ends up. Would you happen to know off-hand?

Edit: Also, thanks. I'd forgotten about this oddity. I believe it's related to how `_log_invocation` works versus how `_handle_no_log_values` works. A parameter found in `_log_invocation` never makes it into `no_log_values`.

Comment (samdoran) [reply docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: It gets passed to the `log()` method, but it doesn't actually mask anything since `remove_values()` is using the values in `self.list_no_log_values`.

This is probably a bug because I bet the intention was to mask those values if they looked like that contained sensitive information.

Comment (samdoran) [reply docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: Oh, nevermind. It _does_ actually get masked in the log message, just not in the module invocation.

```
syslog_msg='Invoked with password_length=NOT_LOGGING_PASSWOR...t_elements
```

Comment (nnhanndakara) [reply docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: Ahh. That's what I was trying to test. Thanks.

Comment (samdoran) [reply docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: So the value is masked in the syslog, but not the Ansible log/output.

Comment (nnhanndakara) [reply docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: You could technically add the value to `no_log_values` on the module object with each invocation of `_log_invocation` (since it's a `set`) and eat the cost of the call, but that would  result in an inconsistency between the resulting messages, I believe. `"NOT_LOGGING_PASSWORD"` in syslog, and `"NO_LOGGING_PARAMETER"` in the output.

Comment (nnhanndakara) [reply docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: For the purposes of this PR, should we be documenting the intended behavior, or the current behavior, including the bug?

Comment (samdoran) [reply docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: I wonder if that's an omission. I will ask around. It's simple to also mask that value in output.

```patch
diff --git lib/ansible/module_utils/basic.py lib/ansible/module_utils/basic.py
index a96ce6aa5d..14a22bdf59 100644
--- lib/ansible/module_utils/basic.py
+++ lib/ansible/module_utils/basic.py
@@ -1948,6 +1948,7 @@ class AnsibleModule(object):
             elif PASSWORD_MATCH.search(param) and arg_opts.get('type', 'str') != 'bool' and not arg_opts.get('choices', False):
                 # skip boolean and enums as they are about 'password' state
                 log_args[param] = 'NOT_LOGGING_PASSWORD'
+                self.no_log_values.add(self.params[param])
                 self.warn('Module did not set no_log for %s' % param)
             else:
                 param_val = self.params[param]

```

Comment (samdoran) [reply docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: We discussed this internally, and adding the suspect value to `self.no_log_values` could potentially cause issues by masking a lot of output strings that don't need to be masked.

Comment (nnhanndakara) [reply docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: Potentially. If the consensus is that the responsibility placed on module authors also needs to come with some safeguards then I can understand the logic behind it. Should this be documented that way in the documentation then?

Comment (samdoran) [reply docs/docsite/rst/dev_guide/developing_program_flow_modules.rst]: Yes. It is up to the module author to mark fields as `no_log`.

**Comment by samdoran** [comment]
We have integration tests for `no_log` [here](https://github.com/ansible/ansible/tree/devel/test/integration/targets/no_log). You should be able to add to these tests for this new behavior.

**Review by samdoran — APPROVED** [review]

**Comment by samdoran** [comment]
This is listed as a bug fix, but it is also a change in behavior. I'll ask for other opinions about whether or not it should be backported.

**Comment by kaorihinata** [comment]
Not all modules are tested like `user`. There are at least a few modules currently in the tree that will emit a warning with this change, but they should be pretty easy to spot in CI output.

**Comment by samdoran** [comment]
The consensus on backporting this is not strong either way, so go ahead and open a [backport PR](https://docs.ansible.com/ansible/latest/community/development_process.html#backporting-merged-prs).

**Comment by kaorihinata** [comment]
Should be doable. I'm moving house so it may not happen immediately, but I'll try to get it thrown together between bouts of filling boxes.
